File: Homework6/Hw6/Calculator.fs
module Hw6.Calculator

type CalculatorOperation =
     | Plus = 0
     | Minus = 1
     | Multiply = 2
     | Divide = 3


[<System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage>]
let inline calculate (value1: double, operation , value2: double) =
     match operation with
     | Error message -> Error message
     | Ok oper ->
          match oper with
          | CalculatorOperation.Plus -> Ok $"{value1 + value2}"
          | CalculatorOperation.Minus -> Ok $"{value1 - value2}"
          | CalculatorOperation.Multiply -> Ok $"{value1 * value2}"
          | CalculatorOperation.Divide ->
               if value2 = 0.0 then
                    Error "Error! Dividing by zero!"
               else
                    Ok $"{value1 / value2}"

    

File: Homework8/Hw8/Calculator/Calculator.cs
namespace Hw8.Calculator;

public class Calculator: ICalculator
{
    public double Calculate(CalculatorArgs args)
    {
        return args.Operation switch
        {
            Operation.Plus => Plus(args.Value1, args.Value2),
            Operation.Minus => Minus(args.Value1, args.Value2),
            Operation.Multiply => Multiply(args.Value1, args.Value2),
            Operation.Divide => Divide(args.Value1, args.Value2),
            _ => throw new InvalidOperationException(Messages.InvalidOperationMessage)
        };
    }
    public double Plus(double val1, double val2) => val1 + val2;
    
    public double Minus(double val1, double val2) => val1 - val2;
    
    public double Multiply(double val1, double val2) => val1 * val2;

    public double Divide(double firstValue, double secondValue)
    {
        if (secondValue == 0)
            throw new ArgumentException(Messages.DivisionByZeroMessage);

        return firstValue / secondValue;
    }
}